yarn-project/simulator/src/avm/opcodes/instruction.ts

```diff
@@ -14,14 +15,32 @@ type InstructionConstructor = {
  * It's most important aspects are execute and (de)serialize.
  */
 export abstract class Instruction {
+  /**
+   * Consumes gas and executes the instruction.
+   * This is the main entry point for the instruction.
+   * @param context - The AvmContext in which the instruction executes.
+   */
+  public run(context: AvmContext): Promise<void> {
+    context.machineState.consumeGas(this.gasCost());
+    return this.execute(context);
+  }
+
+  /**
+   * Loads default gas cost for the instruction from the GasCosts table.
+   * Instruction sub-classes can override this if their gas cost is not fixed.
+   */
+  public gasCost(): GasCost {
+    return GasCosts[this.opcode] ?? EmptyGasCost;
```

Comment:
Can we make this fail statically if `GasCosts[this.opcode]` is undefined?

Reply:
Since `GasCosts: Record<Opcode, GasCost>`, it should fail statically if `GasCosts` is missing an `Opcode`. And since `public get opcode(): Opcode`, we should be able to safely remove the `?? EmptyGasGost`.

Reply:
Yep, @just-mitch you're right!